README.md
![logo](http://linkeddata.center/resources/v4/logo/Logo-colori-trasp_oriz-640x220.png)
# BOTK\Core
[![Build Status](https://img.shields.io/travis/linkeddatacenter/BOTK-core.svg?style=flat-square)](http://travis-ci.com/linkeddatacenter/BOTK-core)
[![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/linkeddatacenter/BOTK-core/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/linkeddatacenter/BOTK-core/?branch=master)
[![Code Coverage](https://img.shields.io/scrutinizer/coverage/g/linkeddatacenter/BOTK-core.svg?style=flat-square)](https://scrutinizer-ci.com/g/linkeddatacenter/BOTK-core)
[![Latest Version](https://img.shields.io/packagist/v/botk/core.svg?style=flat-square)](https://packagist.org/packages/botk/core)
[![Total Downloads](https://img.shields.io/packagist/dt/botk/core.svg?style=flat-square)](https://packagist.org/packages/botk/core)
[![License](https://img.shields.io/packagist/l/botk/core.svg?style=flat-square)](https://packagist.org/packages/botk/core)

Super lightweight base classes for developing smart gateways to populate RDF knowlege base.


## Installation

The package is available on [Packagist](https://packagist.org/packages/botk/core).
You can install it using [Composer](http://getcomposer.org):

	composer require botk/core

## Overview

This package provides some simple tools to transform  raw data into rdf linked data.

This package is compatible with [LinkedData.Center SDaaS architecture](http://linkeddata.center/home/sdaas)

It provides:

- a set of libraries to help gateways development
- a set of libraries to help reasoners development



## Libraries usage

The goal of the libraries is to simplify the conversion of raw data (e.g. .csv or  xml file) in RDF. There are tons of tools to do this job and this is yet another. The idea is just to use PHP to do trivial data conversion instead to build and configure complex tool.

With BOTK you define simple models to describe things  with  a plain set of properties ( e.g a Business Entity, a contact, a product). Then a FactFactory processor cleans data and translates attributes in a RDF graph according with BOTK language profile.
More or less this is what you have to do do when process csv or excel files row by row.

With BOTK libraries it is easy to create "gateways" ie processors that get in stdin a data stream producing in sdout a RDF turtle stream

For example this code snippet:

```php
$options = [
    'factsProfile' => [
        'model' => 'SampleSchemaThing',
        'modelOptions' => [
            'base' => [ 'default'=> 'urn:yp:registry:' ]
        ],
        'datamapper'	=> function($rawdata){
            $data = array();
            $data['identifier'] = $rawdata[0];
            $data['homepage'] =  $rawdata[1];
            $data['alternateName'] = [ $rawdata[2], $rawdata[3]] ;
            return $data;
        },
        'rawdataSanitizer' => function( $rawdata){
            return (count($rawdata)==4)?$rawdata:false;
        },
     ],
    'skippFirstLine'	=> true,
    'fieldDelimiter' => ','
];

BOTK\SimpleCsvGateway::factory($options)->run();
```

processes this csv dataset:

	 id,url,name,aka 
	 1,http://linkeddata.center/,LinkedData.Center,LDC
	 2,https://github.org/,GitHub,


and produces something similar to this RDF turtle file:
	
	@prefix schema: <http://schema.org/> .
	
	<urn:yp:registry:1> 
		schema:alternateName "LDC","LinkedData.Center" ;
	    schema:url <http://linkeddata.center/> .
	
	<urn:yp:registry:2> 
		schema:alternateName "GitHub" ;
	    schema:url <https://github.org/> .


The the dataset processing is driven by the SimpleCsvGateway class that uses a set of options that you can override:

| SimpleCsvGateway option | default   | note                                         |
|-------------------------|-----------|----------------------------------------------|
| missingFactsIsError     | true      | if a missing fact should considered an error |
| bufferSize              | 2000      |                                              |
| skippFirstLine          | true      |                                              |
| fieldDelimiter          | ','       |                                              |
| factsProfile            | array ... | see below                                    |

**factsProfile** are processed by FactsFactory class that uses following options:

| factsProfile option | default               | note                                                                                                              |
|---------------------|-----------------------|-------------------------------------------------------------------------------------------------------------------|
| model               | LocalBusiness         | if no namespace specified BOTK\Model is used                                                                      |
| modelOptions        | array ...             | see below                                                                                                         |
| entityThreshold     | 100                   | in numbers of entity that trigger error resilence computation                                                     |
| resilienceToErrors  | 0.3                   | if more than 30% of error throws a TooManyErrorException                                                          |
| resilienceToInsanes | 0.9                   | if more than 90% of unacceptable data throws a TooManyErrorException                                              |
| source              | 'http://example.com/' | the dataset source url                                                                                            |
| datamapper          | function              | ** YOU must provide at least this function **                                                                     |
| dataCleaner         | function              | a function to clean fields, by default removes ampty fields                                                       |
| factsErrorDetector  | function              | a function that detects logical errors in facts, by default reurne false (i.e. error) if no rdf triples generated |
| rawdataSanitizer    | function              | a function that pre-validate raw data before processing, can be used as a filter                                  |


**modelOptions**  override the default field options provided by the selected model in the $DEFAULT_OPTIONS variable. 
For example see this code snippet extracted from [Thing model](src\Model\Thing.php) that is a superclass of [LocalBusiness model](src\Model\LocalBusiness.php)

Configuring models Options you can force field clenacing and validation.

```php
...
	'uri' => array(
		'filter'    => FILTER_CALLBACK,
		'options' 	=> '\BOTK\Filters::FILTER_VALIDATE_URI',
		'flags'  	=> FILTER_REQUIRE_SCALAR
	),
	'base' => array(
		'default'	=> 'urn:local:',
		'filter'    => FILTER_CALLBACK,
		'options' 	=> '\BOTK\Filters::FILTER_VALIDATE_URI',
		'flags'  	=> FILTER_REQUIRE_SCALAR
	),
	'id' => array(
		'filter'    => FILTER_CALLBACK,
		'options' 	=> '\BOTK\Filters::FILTER_SANITIZE_ID',
		'flags'  	=> FILTER_REQUIRE_SCALAR
	),
	'page' => array(	
		'filter'    => FILTER_CALLBACK,
		'options' 	=> '\BOTK\Filters::FILTER_SANITIZE_HTTP_URL',
		'flags'  	=> FILTER_FORCE_ARRAY
	),
...
```

a field definition drives the process of data cleansing and rdf generation that is provided by model implementation.
Note that not always a field  generate just a RDF triple: sometime the rdf generation processing requires to create blank nodes or to reference named node.
For named node generation the 'base' uri namespace is normally used ("urn:local:." by default)

See [more examples here](tests/system/gateways).


## Contributing to this project

See [Contributing guidelines](CONTRIBUTING.md)

## License

Copyright © 2018-2021 by [LinkedData.Center](http://LinkedData.Center/)®

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/) and [keepchangelog format](http://keepachangelog.com/)
From 9.0.0, this project adheres to the [conventional commit specification](https://www.conventionalcommits.org/en/v1.0.0/)


## [Unreleased]

## 9.0.0

### Added
- docs: adopted conventional commits
- 
### Changed
- feat(composer)!: requires PHP >= 8.2 
- feat(composer)!: requires phpunit 10
- fix(tests): unit tests compatibility

## 8.3.0

### added

- 'dataInStream'  and 'dataOutStream' new options in SimpleCsvGateway
- SimpleCsvGateway and FactsFactory.factualize accept an extra argument 
  for a storage object to be used to store data visible to the with caller stack
- SimpleCsvGateway unit test

## 8.2.5

### added

- compatibility with php 8 ( issue #19)

### removed

- obsolete system tests and their dependencies with old SDaaS platform versions


## 8.2.4

### fixed
- regression
- bug #12 #13 #15

### added

- added 'dct', 'void', and 'foaf' prefix to the Abstract Model
- documentUri to options to drive metadata printing

### removed

- provenance info
- system tests with old relase of sdaas

### changed

-  rdf footer


## 8.2.3

### fixed

- regression

## 8.2.2

### fixed

- regression

## 8.2.1

### added

- added 'dct', 'void', and 'foaf' prefix to the Abstract Model
- documentUri to options to drive metadata printing

### fixed

- bug #15

### removed

- provenance info

### changed

- footer

## 8.2.0

### fixed

- bugs #12 #13  

## removed

- system tests with old relase of sdaas

## 8.1.1

### added

- code quality badge
- added more unit tests

### fixed

- fixed default model in FactsFactory (interface bug)
- added getUri suffix optional parameter (interface bug)
- bug in getURI

## 8.1.0

Various bug fixed on v8 interface

### added

- system tests

## 8.0.0

A project re-factory that removes specific ontology model just keeping only the core classes.
Not good for production

### removed

- all models except things
- all unused test
- the language profile

### Changed

- to php 7.3
- to phpunit 9
- AbstractModel includes core $DEFAULT_OPTIONS (uri, base, id)

## 7.6.1


## fixed

- \mb_str.. in filters

## 7.6.0

### Changed

- Substituted vagrat with docker


### addded

- Abstract reasoner

## removed

- command line interface

## 7.5.1

### addded

- factsFactory as conatiner
- file argument

## 7.5.0

### added

- normalized quantitative value axioms
- botk application and postman	:reasoning command

## fixed

- botk:similarTo property

## removed

- botk:similarName propery

## 7.4.2

### added

- model code and properties for implementing Italian Postal code


## 7.4.1

### fixed

- email in LocalBusiness is string not an uri.


### changed

- readme fix

## 7.4.0 update language profile

### added

- added administrative area definition for italy

## 7.3.7 bug fixing release

### fixed

- bug #10



## 7.3.6 bug fixing release

### fixed

- fixed failing test


## 7.3.6 bug fixing release

### fixed

- fixed failing test


## 7.3.5 bug fixing release

### fixed

- bug #9 (taxID)

## 7.3.4 bug fixing release

## 7.3.3 bug fixing release

### fixed

- bugs in language profile documentation and Readme

## removed

- alternateNames.constuct and linkToCityConstruct


## 7.3.2 bug fixing release

### fixed

- bugs in language profile documentation


## 7.3.1 bug fixing release

### fixed

- bug in local_business_answers.ttl
- bug in linkCities.construct
- bugs in some queries


## 7.3.0

### Fixed

- sameas moved to sameAs in Thing


### Added

- silent option

 
## 7.2.0


### Changed

- README.md improvements

### Added

- restored CONTRIBUTING.md file
- example axioms
 

## 7.1.0

### Added

- property foundingDate in LocalBusiness
 
### Changed

- fixed cardinality errors in LocalBusiness
- refactory of rdf generation in BusinessContact
- fixed vocabulary description


## 7.0.2

### Changed

- minor bug fixing and code opimization

## 7.0.1

### Changed

- Bug in trevis config

## 7.0.0

### Changed

- Code refactory
- optimization in rdf generation
- minor changes to restrictions in BOTK language profile

### Added

- Product and Thing model 
- added sample6.php
- functional tests
- raptor installation in vagrant


## 6.4.0

### Added

- BusinessContact model with lot of new properties
- added sample5B.php


### Changed

- doc: CONTRIBUTING.MD merged with README.
- composer updated.

## 6.3.0


### Added

- added lot of new properties
- added sample5A.php

### Changed

- LocalBusiness rdf creation strategy

## 6.2.1

### Added

- added new properties: annualTurnover, netProfit, ateco2007 and EBITDA

## 6.2.0

### Added

- added new properties numberOfEmployees

### Changed
- changed botk namespace

### Fixed

- fixed error in entity count in FactsFactory

## 6.1.0

### Added

- addedd property numberOfEmployees

### Changed

- Refactory.
- Changed Model interface

## 6.0.0

Completelly rewitten

## 5.1.1 - 2016-01-29

### Fixed

- bug in AbstractCOntentNegotiation

## 5.1.0 - 2016-01-29

### Changed

- refactory of Standard Representation

## 5.0.1 - 2016-01-29
### Fixed

- removed install script problem

### Changed

- removed subversion dependency
- some project cleaning

## 5.0.0 - 2016-01-28

### Fixed
- html renderer bugs

### Removed

- removed xml renderer support (who care of a genetic xml application?)

### Changed

- moved default css to linkeddata.center resorces
- php unit refactory

### Added

- vagrant support
- contribution instructions
- badge management
- trevis-ci support
- scrutinizer support


## 4.0.1 - 2016-01-28

### Added

- integration with packagist
- changelog alignement

## 4.0.0 - 2016-01-28

### Added
- Moved from google code
- Code, doc and tests completed


src/FactsFactory.php
<?php
namespace BOTK;

use BOTK\Exception\FactsValidatorWarning;
use BOTK\Exception\InvalidRawDataWarning;
use BOTK\Exception\TooManyErrorsException;
use BOTK\Exception\TooManyInsanesException;

/**
 * Create structured data from an array of raw data (i.e. ie just a sequence of scalars) managing errors.
 * The class provides a RDF triple counter and tresholds for data processinge errors
 * Options:
 * 	'datamapper' a callable that accepts a raw data array and create structured data as an array. Must be provided.
 *  'rawdataSanitizer a callable that validate raw data before datamapper. It returns an array of raw data of false if rawdata is invalid.
 *  'dataCleaner' a callable that filters the structured data returned by datamapper before to instanciate data model, 
 * 					by default it removes all empty properties.
 *  'factsErrorDetector' a callable that validate computed facts. It accepts a ModelInterface and returns an error description  or false. 
 * 					By defaults accepted raw data that produces empty facts are considered errors.
 */
class FactsFactory implements FactsFactoryInterface {
	
	protected $profile;
	protected $modelClass;
	protected $counter = array(
		'triple'		=> 0,			// rdf triples in facts
		'error'			=> 0,			// facts contains error
		'insane'		=> 0,			// raw data unaccepted
		'entity'		=> 0,			// raw data processed
	);
	
	
	public function __construct( array $profile =array() )
	{
		$defaults = array(
			'model'					  => 'SampleSchemaThing',
			'modelOptions'			  => array(),
			'entityThreshold'		  => 100, // min numbers of entity that trigger error resilence computation.
			'resilienceToErrors' 	  => 0.3, // if more than 30% of error throws a TooManyErrorException
		    'resilienceToInsanes'	  => 0.9, // if more than 90% of unacceptable data throws a TooManyErrorException
		    'documentURL' 	          => null,	// used if printing metadata to specify source file URL use empty string for <>
		    'source' 			  	  => null,
			'datamapper'			  => function($rawdata){return $rawdata;},
			'dataCleaner' 		  	  => get_class().'::REMOVE_EMPTY',
			'factsErrorDetector' 	  => get_class().'::NOT_EMPTY_FACTS',
			'rawdataSanitizer' 		  => function($rawdata){return is_array($rawdata)?$rawdata:false;},
		);
		$this->profile = array_merge($defaults,$profile);
		$this->modelClass = class_exists($this->profile['model'])
			?$this->profile['model']
			:('\BOTK\Model\\'.$this->profile['model']);
		
		if( !class_exists($this->modelClass) || !is_subclass_of($this->modelClass, '\BOTK\ModelInterface')){
			throw new \InvalidArgumentException("The provided model ({$this->profile['model']} is unknown");	
		}
		if( !is_callable($this->profile['datamapper'])) {
			throw new \InvalidArgumentException("Invalid datamapper callback");	
		}
		if( !is_callable($this->profile['dataCleaner'])) {
			throw new \InvalidArgumentException("Invalid dataCleaner callback");	
		}
		if( !is_callable($this->profile['rawdataSanitizer'])) {
			throw new \InvalidArgumentException("Invalid rawdataSanitizer callback");	
		}
		if( !is_callable($this->profile['factsErrorDetector'])) {
			throw new \InvalidArgumentException("Invalid factsErrorDetector callback");	
		}
	}
	
	
	/**
	 * two level filter array, a default for dataCleaner callback
	 */
	public static function REMOVE_EMPTY( array $data)
	{
		$a = array();
	    foreach ($data as $key => $value) {
	       $a[$key] = is_array($value)?array_filter($value):$value;
	    }
	    return array_filter($a);
	}
	
	
	/**
	 * a default for dataValidator callback  
	 */
	public static function NOT_EMPTY_FACTS( \BOTK\ModelInterface $data)
	{
		return $data->getTripleCount()?false:'No facts found.';
	}
	
	
	/**
	 * create facts from rawdata. Please nothe that null facts does not means always an error (i.e. no facts is a fact).
	 * if you do not want empty facts use dataValidator
	 */
	public function factualize($rawData,  object $globalStorage=null)
	{
	    
		$rawdataSanitizer = $this->profile['rawdataSanitizer'];
		$validRawData = $rawdataSanitizer($rawData);
		$this->counter['entity']++;
		
		if (!empty($validRawData)){
			$datamapper = $this->profile['datamapper'];
			$dataCleaner = $this->profile['dataCleaner'];
			$factsErrorDetector = $this->profile['factsErrorDetector'];
			$data =$dataCleaner($datamapper($validRawData));
			$facts = call_user_func($this->modelClass.'::fromArray', $data, $this->profile['modelOptions'], $globalStorage);
			$this->counter['triple'] += $facts->getTripleCount();
			if($error=$factsErrorDetector($facts)){
				$this->counter['error']++;
				throw new FactsValidatorWarning($error,$facts);
			}
		} else {
			$this->counter['insane']++;
			throw new InvalidRawDataWarning("Invalid rawdata",$rawData);
		}
		
		// ensure that not too many errors
		$errorRate = ($this->counter['error']/$this->counter['entity']);
		if(( $this->counter['entity'] > $this->profile['entityThreshold']) 
				&& ( $errorRate > $this->profile['resilienceToErrors'])){
			$x = $this->profile['resilienceToErrors']*100;
			throw new TooManyErrorsException("Error rate in data processing exceeded the $x% threshold");			
		}

		// ensure that not too many insaness raw data
		$insaneRate = ($this->counter['insane']/$this->counter['entity']);
		if(( $this->counter['entity'] > $this->profile['entityThreshold']) 
				&& ($insaneRate > $this->profile['resilienceToInsanes'])){
			$x = $this->profile['resilienceToInsanes']*100;
			throw new TooManyInsanesException("Unacceptable data rate exceeded the $x% threshold");			
		}

		return $facts;
	}
	
	
	public function generateLinkedDataHeader()
	{
	    $metadata='';
	    if(!is_null($this->profile['documentURL'])) {
	        // Requires foaf:,  dct: and void: prefixes to be defined in model
	        $metadata.= "\n<{$this->profile['documentURL']}> a foaf:Document .\n";
	        if(!empty($this->profile['source'])){
	            $metadata .= "<{$this->profile['documentURL']}> dct:source <{$this->profile['source']}>. \n";
	        }
	        $metadata .= "[] a void:Dataset; dct:format \"text/turtle\" ; void:datadump <{$this->profile['documentURL']}>;void:triples {$this->counter['triple']} ;void:entities {$this->counter['entity']}.\n";	        
	    }
	    $base = $this->profile['modelOptions']['base']['default']?? null;
	    return call_user_func($this->modelClass.'::getTurtleHeader',$base) . $metadata;
	}
	
	
	public function generateLinkedDataFooter()
	{
	    return "#Generated {$this->counter['triple']} good triples from {$this->counter['entity']} entities ({$this->counter['insane']} ignored), {$this->counter['error']} errors\n";
	}


	public function addToCounter($counter,$val=1)
	{
		if(!array_key_exists($counter,$this->counter)){
			throw new \InvalidArgumentException("Invalid counter name");
		}
		$this->counter[$counter]+= intval($val);
	}
	
	
	public function getCounters()
	{
		return $this->counter;
	}

}